File: Stack/ApplicationOfStack/infixToPostfix.cc
#include<bits/stdc++.h>
using namespace std;

// Function to return precedence
int precedence(char c) {

    if(c == '+' || c == '-')
    return 1;
    else if(c == '*' || c == '/')
    return 2;
    else if(c == '^')
    return 3;
    else
    return -1;
    
}

// main function to convert
void converter(string s) {

    std::stack<char> st;
    st.push('N');
    int l = s.length();
    string ns;
    for (int i = 0; i < l; i++){

        // if the scanned element is an operand
        if ((s[i] >= 'a' && s[i] <= 'z') ||
            (s[i] >= 'A' && s[i] <= 'Z'))
            ns += s[i];

        // if the scanned element is open bracket "("
        else if (s[i] == '(')
        st.push('(');

        // if the scanned element is close bracket ")"
        else if (s[i] == ')') {
            while (st.top() != 'N' && st.top() != '(') {
                	
                    char c = st.top();
                    st.pop();
                    ns += c;
            }
            if (st.top() == '(') {
                char c = st.top();
                st.pop();
            }
        }

        // if the scanned element is operator
        else {
            while (st.top() != 'N' && precedence(s[i]) <= precedence(st.top())) {
                char c = st.top();
                st.pop();
                ns += c;
            }
            st.push(s[i]);
        }
    }

    // pop all the remaining elements from stack
    while (st.top() != 'N') {
        char c = st.top();
        st.pop();
        ns += c;
    }
    cout << ns << endl;
}

// function calling
int main() {
    string exp = "a+b*(c^d-e)^(f+g*h)-i";
    converter(exp);
    return 0;
}

File: Stack/ApplicationOfStack/infixToPostfix.java
package Stack.ApplicationOfStack;

import java.util.*;

public class infixToPostfix {
    
    // method to return operator precedence
    static int precedence(char ch) {
        
        switch (ch) {
            
            case '+':
            case '-':
                return 1;
            
            case '*':
            case '/':
                return 2;
            
            case '^':
                return 3;
        }
        return -1;
    }

    /* Method that Converts */
    static String converter(String exp) {

        // initializing string for output
        String result = new String("");

        // initializing stack for operators
        Stack<Character> stack = new Stack<>();

        for (int i = 0; i < exp.length(); i++) {

            char c = exp.charAt(i);

            // if the element is operand
            if (Character.isLetterOrDigit(c))
                result += c;

            // if the element is open bracket '('
            else if (c == '(')
                stack.push(c);
            
            // if the element is close brakcet ')'
            else if (c == ')') {

                while (!stack.isEmpty() && stack.peek() != '(')
                    result += stack.pop();

                stack.pop();
            }
            
            // if an operator is encountered
            else {
                while (!stack.isEmpty() && precedence(c) <= precedence(stack.peek())) {
                    result += stack.pop();
                }
                stack.push(c);
            }
        }

        // pop the remaining operators in the stack
        while (!stack.isEmpty()) {
            if (stack.peek() == '(')
                return "Not a valid expression";
            result += stack.pop();
        }
        return result;
    }

    public static void main(String[] args) {
        String exp = "a*b(c^d-e)^(f+g*h)-i";
        System.out.println(converter(exp));
    }
    
}